cloud-azure/src/main/resources/templates/arm-v2.ftl

```diff
@@ -254,20 +254,28 @@
                        <#if !noFirewallRules || !noPublicIp>,</#if>
                        "[concat('Microsoft.Network/virtualNetworks/', parameters('virtualNetworkNamePrefix'))]"
                        </#if>
+                       <#if instanceGroup == "GATEWAY" && loadBalancers?? && (loadBalancers?size > 0)>,
```

Comment:
This is adding all instances in the `GATEWAY` instance group to the load balancer backend address pool.
In this case `instanceGroup` is the key in a map between the instance group name and the instance group object, so it's actually a string, one of `GATEWAY`, or `CORE`.

This line is inside two nested loops (pseudocode):
```
for each instanceGroup
    for each instance in the instanceGroup
        if instanceGroup == gateway then add to backend address pool
```

Which is a consequence of how we have to associate the NICs with a backend address pool.
We can either:
* Provision the NICs with a `backendAddressPool` **id** to define their membership
* Provision the `backendAddressPool` with the **ip addresses** of the NICs that belong to it

It was easier to do the first option. There are some issues with this set up, namely that we'll need to fix how load balancers are associated with instance groups generally as we add more load balancers.

It feels like the "instanceGroups" abstractions has some common properties that should be crystallized as we beef up the DL shape:
* An instance group has an availability set
* It has it's own load balancer
* It should be distributed across fault domains
* It should be distributed across AZs

So, I was hoping to include fixes for this load balancer logic alongside some other Azure improvements as we move forward.

Reply:
> * It has it's own load balancer

This isn't necessarily true for AWS. From what I remember of our discussions, the Azure  LB needs to be in the same availability set as the instances its routing to. And you've made it so that each instance group is in its own availability set. So that all makes sense for Azure. But the thing is, while in all the clusters we've seen so far Knox is installed on the gateway instances, there's no programmatic requirement to that effect. CM is always installed on the gateway instances. But Knox can theoretically be installed in any instance group, or on multiple instance groups.

To deal with this possibility, the `LoadBalancerConfigService` iterates over the instance groups and finds all the ones that have Knox installed. It then sets up the connection between the `TargetGroup` entity and the `InstanceGroup` entity, where a instance group can have multiple target groups (as in the semi-private case where both the public and private LB are routing traffic to the same targets), and the target group can have multiple instances groups, for the theoretical case where Knox is installed on multiple instance groups. On the cloud provider side, this is represented in the `CloudLoadBalancer.portToTargetGroupMapping` map by having the `TargetPortPair` key point to a _set_ of `Group`.

So this is all a long winded way to say that in order to be perfectly safe for all supported configurations, we shouldn't have a hardcoded check for the type Instead we should check the name of the instance group, and see if it's in the set of instance groups the LB is supposed to route to.

Reply:
Ok! This makes sense.
Looks like I can map the target groups to azure load balancing rules pretty directly.

Getting the VM instances correctly associated with the backend address pools is a little tricky. I think I'm going to have to do something gross in the ftl file:
```
"loadBalancerBackendAddressPools": [
   <#list loadBalancers as loadBalancer>
       <#list loadBalancer.rules as rule>
           <#list rule.backendPool.backendInstances as candidate>
               <#if instance.instanceId == candidate.instanceId>
               {
                   "id": "[resourceId('Microsoft.Network/loadBalancers/backendAddressPools', '${rule.backendPool.name}')]"
               }
               </#if>
           <#sep>, </#list>
       <#sep>, </#list>
   <#sep>, </#list>
]
```

Reply:
Following up with what Bryce and I decided on. This case is difficult to handle in Azure because:

- Availability sets are currently configured such that each instance group is in its own availability set.
- Azure load balancers with the Basic SKU can only route traffic to instances they share an availability set with.

That means that there's no way, short of very messy handling of availability sets, to handle Knox being installed on multiple instance groups. So instead, we've introduced a condition in `LoadBalancerConfigService` that will check is there's more than one Knox target instance group for Azure, and will not create the load balancer objects if so.

Once we move to using the Azure Standard SKU generally the availability set restriction will no longer apply, and we can remove that restriction in the code.